File: projectBlueWater/src/main/java/com/lasthopesoftware/bluewater/client/connection/builder/specs/GivenAnAccessCodeThatIsAValidServer/WhenScanningForUrls.java
package com.lasthopesoftware.bluewater.client.connection.builder.specs.GivenAnAccessCodeThatIsAValidServer;

import com.lasthopesoftware.bluewater.client.browsing.library.repository.Library;
import com.lasthopesoftware.bluewater.client.connection.builder.UrlScanner;
import com.lasthopesoftware.bluewater.client.connection.builder.lookup.LookupServers;
import com.lasthopesoftware.bluewater.client.connection.okhttp.OkHttpFactory;
import com.lasthopesoftware.bluewater.client.connection.testing.TestConnections;
import com.lasthopesoftware.bluewater.client.connection.url.IUrlProvider;
import com.lasthopesoftware.bluewater.shared.promises.extensions.specs.FuturePromise;
import com.lasthopesoftware.resources.strings.EncodeToBase64;
import com.namehillsoftware.handoff.promises.Promise;

import org.junit.BeforeClass;
import org.junit.Test;

import java.util.concurrent.ExecutionException;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.ArgumentMatchers.argThat;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

public class WhenScanningForUrls {

	private static IUrlProvider urlProvider;

	@BeforeClass
	public static void before() throws InterruptedException, ExecutionException {
		final TestConnections connectionTester = mock(TestConnections.class);
		when(connectionTester.promiseIsConnectionPossible(argThat(a -> a.urlProvider.getBaseUrl().equals("http://gooPc:80/MCWS/v1/"))))
			.thenReturn(new Promise<>(true));

		final UrlScanner urlScanner = new UrlScanner(
			mock(EncodeToBase64.class),
			connectionTester,
			mock(LookupServers.class),
			OkHttpFactory.getInstance());

		urlProvider = new FuturePromise<>(
			urlScanner.promiseBuiltUrlProvider(new Library()
				.setAccessCode("gooPc"))).get();
	}

	@Test
	public void thenTheUrlProviderIsReturned() {
		assertThat(urlProvider).isNotNull();
	}

	@Test
	public void thenTheBaseUrlIsCorrect() {
		assertThat(urlProvider.getBaseUrl()).isEqualTo("http://gooPc:80/MCWS/v1/");
	}
}

File: projectBlueWater/src/test/java/com/lasthopesoftware/bluewater/client/connection/builder/GivenANullAccessCode/WhenScanningForUrls.kt
package com.lasthopesoftware.bluewater.client.connection.builder.GivenANullAccessCode

import com.lasthopesoftware.bluewater.client.browsing.library.repository.LibraryId
import com.lasthopesoftware.bluewater.client.connection.builder.UrlScanner
import com.lasthopesoftware.bluewater.client.connection.settings.LookupConnectionSettings
import com.lasthopesoftware.bluewater.client.connection.url.IUrlProvider
import com.lasthopesoftware.bluewater.shared.promises.extensions.toFuture
import com.namehillsoftware.handoff.promises.Promise
import io.mockk.every
import io.mockk.mockk
import org.assertj.core.api.Assertions.assertThat
import org.junit.BeforeClass
import org.junit.Test
import java.util.concurrent.ExecutionException

class WhenScanningForUrls {

	companion object {
		private var illegalArgumentException: IllegalArgumentException? = null
		private var urlProvider: IUrlProvider? = null

		@BeforeClass
		@JvmStatic
		fun before() {
			val connectionSettingsLookup = mockk<LookupConnectionSettings>()
			every { connectionSettingsLookup.lookupConnectionSettings(any()) } returns Promise.empty()

			val urlScanner = UrlScanner(
				mockk(),
				mockk(),
				mockk(),
				connectionSettingsLookup,
				mockk()
			)
			try {
				urlProvider = urlScanner.promiseBuiltUrlProvider(LibraryId(32)).toFuture().get()
			} catch (e: ExecutionException) {
				if (e.cause is IllegalArgumentException) illegalArgumentException =
					e.cause as IllegalArgumentException? else throw e
			}
		}
	}

	@Test
	fun thenANullUrlProviderIsReturned() {
		assertThat(urlProvider).isNull()
	}

	@Test
	fun thenAnIllegalArgumentExceptionIsThrown() {
		assertThat(illegalArgumentException).isNotNull
	}

	@Test
	fun thenTheExceptionMentionsTheLibrary() {
		assertThat(illegalArgumentException?.message).isEqualTo("The access code cannot be null")
	}
}
